yasa/hypno.py

```diff
@@ -474,3 +477,231 @@ def hypno_find_periods(hypno, sf_hypno, threshold="5min", equal_length=False):
+
+    Parameters
+    ----------
+    hypno : array_like
+        The sleep staging (hypnogram) 1D array.
+    n_stages_in : int
+        Number of possible stages of ``hypno``, where:
+
+        - 5 stages - 0=Wake, 1=N1, 2=N2, 3=N3, 4=REM
+        - 4 stages - 0=Wake, 1=Light, 2=Deep, 3=REM
+        - 3 stages - 0=Wake, 1=NREM, 2=REM
+        - 2 stages - 0=Wake, 1=Sleep
+
+        .. note:: The default YASA values for Unscored (-2) and Artefact (-1) are always allowed.
+    n_stages_out : int
+        Similar to ``n_stages_in`` but for output. Must be higher than ``n_stages_out``.
+
+    Returns
+    -------
+    hypno : array_like
+        The hypnogram, with stages converted to ``n_stages_out`` staging scheme.
+    """
+    assert isinstance(hypno, (list, np.ndarray, pd.Series)), "hypno must be array_like"
+    hypno = np.asarray(hypno, dtype=int).copy()
+    assert n_stages_in in [3, 4, 5], "n_stages_in must be 3, 4, or 5"
+    assert n_stages_out in [2, 3, 4], "n_stages_in must be 2, 3, or 4"
+    assert n_stages_out < n_stages_in, "n_stages_out must be lower than n_stages_in"
+    assert all(s < n_stages_in for s in hypno), "hypno and n_stages_in are not compatible"
+
+    # Change sleep codes (Wake, Artefact, and Unscored never change)
+    if n_stages_out == 2:
+        # Reduce to Wake/Sleep
+        hypno[hypno > 0] = 1  # All non-Wake becomes Sleep
+    elif n_stages_out == 3:
+        # Reduce to Wake/NREM/REM
+        if n_stages_in == 4:  # input is Wake/Light/Deep/REM
+            hypno[hypno == 2] = 1  # Deep become NREM
+            hypno[hypno == 3] = 2  # REM code changes
+        elif n_stages_in == 5:  # input is Wake/N1/N2/N3/REM
+            hypno[hypno == 2] = 1  # N2 becomes NREM
+            hypno[hypno == 3] = 1  # N3 becomes NREM
+            hypno[hypno == 4] = 2  # REM code changes
+    elif n_stages_out == 4:
+        # Reduce to Wake/Light/Deep/REM
+        hypno[hypno == 2] = 1  # N2 becomes Light
+        hypno[hypno == 3] = 2  # N3 becomes Deep
+        hypno[hypno == 4] = 3  # REM code changes
+    return hypno
+
+
+#############################################################################
+# SIMULATION
+#############################################################################
+
+
+def simulate_hypno(tib=90, sf=1 / 30, n_stages=5, trans_probas=None, init_probas=None, seed=None):
+    """Simulate a hypnogram based on transition probabilities.
+
+    Current implentation is a naive Markov model. The initial stage of a hypnogram
+    is generated using probabilites from ``init_probas`` and then subsequent stages
+    are generated from a Markov sequence based on ``trans_probas``.
+
+    .. important:: The Markov simulation model is not meant to accurately portray sleep
+        macroarchitecture and should only be used for testing or other unique purposes.
+
+    Parameters
+    ----------
+    tib : int
+        Time in bed, expressed in minutes.
```

Comment:
Maybe add "total duration of the hypnogram" just to be extra clear?

Reply:
Added this and also a clarification about how `np.floor` is used to find number of _complete_ epochs.
```
tib : int
    Total duration of the hypnogram (i.e., time in bed), expressed in minutes.
    Returned hypnogram will be slightly shorter if ``tib`` in seconds is not
    evenly divisible by ``sf``.
    .. seealso:: :py:func:`yasa.sleep_statistics`
```

I had some trouble wording that concisely, do you think `tib` should be in seconds instead of minutes? Minutes is simpler on-the-fly but the relationship between seconds and `sf` is more intuitive.

Reply:
The updated wording is great. I vote for TIB in minutes (which is also what's returned by yasa.sleep_statistics)